framework/src/meshgenerators/TransfiniteMeshGenerator.C

```diff
@@ -0,0 +1,542 @@
+
+  // The case when the edge spans quadrants 1 and 4 requires special treament
+  // to periodically switch we compute the angle that needs added to one edge
+  // to identify the entire edge span
+  mooseAssert(x0.norm() > 0.0 && x1.norm() > 0.0,
+              "The point provided cannot generate an arc circle on the edge specified");
+
+  Real arclength = std::acos((x0 * x1) / x0.norm() / x1.norm());
+
+  Real a = std::atan2(x0(1), x0(0));
+  Real b = std::atan2(x1(1), x1(0));
+
+  if (a < 0)
+    a = a + 2 * M_PI;
+  if (std::abs(b - a) > M_PI)
+    b = a + arclength;
+
+  auto it = 0;
+  for (auto rx : param_vec)
+  {
+    Real interval = getMapInterval(rx, 0.0, 1.0, a, b);
+    Real x = P0(0) + rad * std::cos(interval);
+    Real y = P0(1) + rad * std::sin(interval);
+    edge[it] = Point(x, y, 0.0);
+    it++;
+  };
+
+  return edge;
+}
+
+Real
+TransfiniteMeshGenerator::getMapInterval(
+    const Real & xab, const Real & a, const Real & b, const Real & c, const Real & d) const
+// this routine maps a point x\in[a, b] to the corresponding point in the interval [c, d]
+{
+  mooseAssert(std::abs(b - a) > 0.0,
+              "The input interval [a, b] is empty, check that a and b are not identical.");
+  Real xcd = c + (d - c) / (b - a) * (xab - a);
+
+  return xcd;
+}
+
+std::vector<Real>
+TransfiniteMeshGenerator::getPointsDistribution(const Real & edge_length,
+                                                const unsigned int & np,
+                                                const Real & bias) const
+{
+  std::vector<Real> param_vec;
+  Real rx = 0.0;
+
+  if (bias != 1.0)
+  {
+    Real step = 0.0;
+    Real factor = edge_length * (1.0 - std::abs(bias)) / (1.0 - std::pow(std::abs(bias), np - 1));
+    param_vec.push_back(rx);
+    for (unsigned int iter = 1; iter < np; iter++)
+    {
+      step = step + factor * std::pow(bias, Real(iter - 1));
+      rx = getMapInterval(step, 0.0, edge_length, 0.0, 1.0);
+      param_vec.push_back(rx);
+    }
+  }
+  else
+  {
+    const Real interval = edge_length / Real(np - 1);
+    Real rx = 0.0;
+    for (unsigned int iter = 0; iter < np; iter++)
+    {
+      rx = getMapInterval(Real(iter) * interval, 0.0, edge_length, 0.0, 1.0);
+      param_vec.push_back(rx);
+    }
+  }
+  return param_vec;
+}
+
+Real
+TransfiniteMeshGenerator::computeRadius(const Point & P1, const Point & P2, const Point & P3) const
+{
+  Point temp1 = P1 - P2;
+  Real a2 = temp1.norm_sq(); // a is the distance from P1 to P2, but we only it squared
+  Point temp2 = P3 - P1;
+  Real b2 = temp2.norm_sq();
+  Real dr = std::sqrt(b2 - a2 / 4.0);
+  const Real rad = a2 / 8.0 / dr + dr / 2.0;
+  return rad;
+}
+
+Point
+TransfiniteMeshGenerator::computeOrigin(const Point & P1, const Point & P2, const Point & P3) const
+{
+  // define interim quantities
+  Real a1, a2, A;
+  if (MooseUtils::absoluteFuzzyGreaterThan(std::abs(P2(0) - P1(0)), 0.0) ||
+      MooseUtils::absoluteFuzzyGreaterThan(std::abs(P2(1) - P1(1)), 0.0))
```

Comment:
Do we really need to handle the case where the points are basically identical?  How does that come up?

Reply:
If either the x or why y coordinates are identical we need a different treatment thus the "else" statement following. It has to do with the system of equations to be solved. different choices need to be made for different cases.

Reply:
The "else" case isn't always working either.  I was suspicious that I screwed up my hand-calculations, so I finally got around to just compiling this function into a stand-alone executable, and I'm still getting the same results.  `P3(1)==P2(1)` gets us `b2==0` and fails.  Looks like you have a `b2 / b2` in the `y0` equation that could be canceled out?

Reply:
Can you be specific about the test you do to replicate and see. It was a very hairy system to solve and had to be done analytically. I couldn't find a standard formula anywhere, if you know any it would be great. Till then I need to see if there's an edge case I might have missed.

Reply:
Fixed, yes you were right it needed one more expand and simplify. I hope this now is more robust.